### PR TITLE
Removed redundant tests

### PR DESCRIPTION
**Pull request by yaelbh**
Some of the gates are tested twice: first time in files like `qasm_standard_gates`, `statevector_gates`, and `unitary_gates`; and second time in files like `qasm_cliffords`, `qasm_noncliffords`, `statevector_basics`, and `unitary_basics`. This pull request removes tests, such that gates are not double-tested.

### CODE REVIEW

**Review by chriseclectic — CHANGES_REQUESTED** [review]
This is a nice cleanup, however it looks like there are currently no gate tests being run for the stabilizer simulator method after this change.

Since the stabilizer simulator does not support statevector or density matrix comparison we could add a special case where for comparing the stabilizier simulator using a stabilizer snapshot and compare to the output of the `Clifford` class from simulating the circuit.

**Comment by yaelbh** [comment]
It's even a bit more complicated, because `qasm_standard_gates` needs to know which gates can run on a Clifford simulator. This can be fixed in the test in several ways, simplest is to just revert the old Clifford tests.

**Comment by yaelbh** [comment]
@chriseclectic Do you know why most sets of basis gates result in this error:
```
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.8/site-packages/ddt.py", line 182, in wrapper
    return func(self, *args, **kwargs)
  File "/home/yaelbh/work/aer/NonCliffordTests/test/terra/backends/qasm_simulator/qasm_standard_gates.py", line 127, in test_gate_statevector
    result = execute(circuit, backend, shots=1, basis_gates=basis_gates,
  File "/home/yaelbh/work/terra/System/qiskit/execute.py", line 252, in execute
    experiments = transpile(experiments,
  File "/home/yaelbh/work/terra/System/qiskit/compiler/transpile.py", line 241, in transpile
    circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_args)))
  File "/home/yaelbh/work/terra/System/qiskit/tools/parallel.py", line 112, in parallel_map
    return [task(values[0], *task_args, **task_kwargs)]
  File "/home/yaelbh/work/terra/System/qiskit/compiler/transpile.py", line 324, in _transpile_circuit
    result = pass_manager.run(circuit, callback=transpile_config['callback'],
  File "/home/yaelbh/work/terra/System/qiskit/transpiler/passmanager.py", line 225, in run
    return self._run_single_circuit(circuits, output_name, callback)
  File "/home/yaelbh/work/terra/System/qiskit/transpiler/passmanager.py", line 288, in _run_single_circuit
    result = running_passmanager.run(circuit, output_name=output_name, callback=callback)
  File "/home/yaelbh/work/terra/System/qiskit/transpiler/runningpassmanager.py", line 113, in run
    dag = self._do_pass(pass_, dag, passset.options)
  File "/home/yaelbh/work/terra/System/qiskit/transpiler/runningpassmanager.py", line 144, in _do_pass
    dag = self._run_this_pass(pass_, dag)
  File "/home/yaelbh/work/terra/System/qiskit/transpiler/runningpassmanager.py", line 156, in _run_this_pass
    new_dag = pass_.run(dag)
  File "/home/yaelbh/work/terra/System/qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py", line 87, in run
    operator = Operator(qc)
  File "/home/yaelbh/work/terra/System/qiskit/quantum_info/operators/operator.py", line 85, in __init__
    self._data = self._init_instruction(data).data
  File "/home/yaelbh/work/terra/System/qiskit/quantum_info/operators/operator.py", line 496, in _init_instruction
    op._append_instruction(instruction)
  File "/home/yaelbh/work/terra/System/qiskit/quantum_info/operators/operator.py", line 554, in _append_instruction
    self._append_instruction(instr, qargs=new_qargs)
  File "/home/yaelbh/work/terra/System/qiskit/quantum_info/operators/operator.py", line 532, in _append_instruction
    raise QiskitError('Cannot apply Instruction: {}'.format(obj.name))
qiskit.exceptions.QiskitError: 'Cannot apply Instruction: snapshot'
```

**Comment by yaelbh** [comment]
Tests are failing because of an issue in Terra:  https://github.com/Qiskit/qiskit-terra/issues/5556
I'll probably set the optimization level to 0.

**Comment by yaelbh** [comment]
Tests take very long to run for some machines, I don't know why, but it makes the PR red.  @vvilpas @mtreinish

**Comment by yaelbh** [comment]
Setting the optimization level to 0 is needed not only as a hack around Terra's bug, but also to make `qasm_standard_gates` really test the identity gate.

**Comment by yaelbh** [comment]
All the tests pass, so ready for review

**Review by vvilpas — COMMENTED** [review]
Code looks good!
Just a doubt, the total number of tests seems to have increased from ~4100 to ~5090 (executed plus skipped). Is that expected? I guess yes, but just checking.

**Comment by yaelbh** [comment]
Yes, it's expected, because every gate is now tested with 5 sets of basis gates. Not sure if these tests are really important, so we can consider removal of some of them.

File: test/terra/backends/qasm_simulator/qasm_standard_gates.py [hunk 1]
```diff
@@ -79,6 +80,14 @@
     (U3Gate, 3),
     (UGate, 3)
 ]
+BASIS_GATES = [
```
Comment (chriseclectic) [inline test/terra/backends/qasm_simulator/qasm_standard_gates.py:83]: Maybe we should update this to just be the old and new IBMQ backend basis gate lists to reduce test load:

```python
BASIS_GATES = [
    None,
    ['id', 'u1', 'u2', 'u3', 'cx'],
    ['id', 'rz', 'sx', 'x', 'cx']
]
```

File: test/terra/backends/statevector_simulator/statevector_gates.py [hunk 2]
```diff
@@ -91,8 +91,8 @@ class StatevectorGateTests:
     ]
     BASIS_GATES = [
```
Comment (chriseclectic) [inline test/terra/backends/statevector_simulator/statevector_gates.py:92]: Same here

File: test/terra/backends/unitary_simulator/unitary_gates.py [hunk 3]
```diff
@@ -91,8 +91,8 @@ class UnitaryGateTests:
     ]
     BASIS_GATES = [
         None,
-        ['id', 'u1', 'u2', 'u3', 'cx'],
-        ['id', 'u', 'cx'],
+        ['id', 'u1', 'u2', 'u3', 'cx'],  # Waltz
```
Comment (chriseclectic) [inline test/terra/backends/unitary_simulator/unitary_gates.py:94]: And here

**Review by chriseclectic — COMMENTED** [review]
Since test timeouts are becoming an issue, we can remove some of the basis gates for the lists, since these are more tests of the transpiler than the simulator. The `basis_gates=None` case should include all supported gates for each method.

**Comment by yaelbh** [comment]
@chriseclectic done

**Review by vvilpas — APPROVED** [review]

**Pull request merged**